WordPress/openverse
# Add catalog indexer worker

## Fixes



Fixes #4147 by @stacimc 

## Description



This PR adds the catalog indexer worker. It does not actually connect to it in the data refresh yet.

Some unit tests are added, but because of the nature of the worker the bulk of the tests will be integration tests modeled after the existing ingestion server integration tests. I was able to get these mostly working locally but pulled them out of this PR as I think it will make more sense to add the integration tests once the data refresh is fully implemented, possibly reusing issue #4340.

Almost all of this code is just a refactor of existing code. Where possible I've noted when files are almost entirely copied over from the ingestion server. 

Note that the service is called `catalog_indexer_worker` to distinguish it from the existing `indexer_worker` that belongs to the ingestion server, which we must keep around for now.

## Testing Instructions


`just build && just up`

Use [elasticvue](https://app.elasticvue.com/cluster/0/indices) to `Delete All Documents` from one of your existing indices. I used `audio-init-filtered`. The index should now have 0 documents.

Now run `just catalog/shell`, and you should be able to curl the indexer worker.

```
# Replace the target_index name if you don't have audio-init-filtered
# Note the start_id/end_id we're using, to make it reindex only 100 records
curl -X POST -d '{"model_name": "audio", "table_name": "audio", "start_id": 0, "end_id": 100, "target_index": "audio-init-filtered"}' -H "Content-Type: application/json" 'http://catalog_indexer_worker:8003/task'
```

You should get back a response like:
```
{
  "message": "Successfully scheduled task.", 
  "task_id": "adbec2cbc93a4c2f895caf12ede3eb88", 
  "status_check": "http://catalog_indexer_worker:8003/task/adbec2cbc93a4c2f895caf12ede3eb88"
}
```

Now curl the status_check endpoint that was just returned:
```
curl 'http://catalog_indexer_worker:8003/task/adbec2cbc93a4c2f895caf12ede3eb88'
```

And you should see something like:

```
{
  "task_id": "adbec2cbc93a4c2f895caf12ede3eb88",
  "active": false,
  "model": "audio",
  "target_index": "audio-init-filtered",
  "progress": 100.0,
  "start_time": "2024-05-14 21:27:54.437745",
  "finish_time": "2024-05-14 21:27:54.575137",
  "error": false
}
```

In elasticvue, you should now see 100 docs in your target index.

If you used a filtered index, you can now query the API (http://localhost:50281/v1/audio) and see that the results are returned as expected. **⚠️ Important: In Elasticvue you must first refresh the index to make the documents available. Then you can inspect them in Elasticvue, or check the API. This will be handled by the DAG later.**

You can also try running the `just` commands for the indexer worker.


## Checklist


      PRs) or the media properties generator (`just catalog/generate-docs media-props`
      for the catalog or `just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
